File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendComplaintRepository.kt
package com.humblesolutions.indsphinx.repository

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.humblesolutions.indsphinx.model.Complaint
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import java.util.Date

class BackendComplaintRepository : ComplaintRepository {
    private val db = FirebaseFirestore.getInstance()

    override suspend fun submitComplaint(complaint: Complaint): String {
        val data = hashMapOf(
            "FlatNumber" to complaint.flatNumber,
            "flatId" to complaint.flatId,
            "OccupantEmail" to complaint.occupantEmail,
            "OccupantName" to complaint.occupantName,
            "OccupantId" to complaint.occupantId,
            "Category" to complaint.category,
            "Date" to Timestamp(Date(complaint.date)),
            "Status" to complaint.status,
            "ResolveDate" to "",
            "Priority" to complaint.priority,
            "Description" to complaint.description,
            "Problem" to complaint.problem,
            "MediaUrls" to complaint.mediaUrls,
            "WorkerName" to complaint.workerName,
            "WorkerUid" to complaint.workerUid,
            "WorkerRemarks" to complaint.workerRemarks,
            "WorkerMedia" to complaint.workerMedia
        )
        val doc = db.collection("Complaints").add(data).await()
        return doc.id
    }

    override suspend fun fetchByOccupant(occupantId: String): List<Complaint> {
        val snapshot = db.collection("Complaints")
            .whereEqualTo("OccupantId", occupantId)
            .orderBy("Date", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.documents.map { doc ->
            val date = (doc.get("Date") as? Timestamp)?.toDate()?.time ?: 0L
            val rd = doc.get("ResolveDate")
            val resolveDate = if (rd is Timestamp) rd.toDate().time else 0L
            Complaint(
                id = doc.id,
                flatNumber = doc.getString("FlatNumber") ?: "",
                flatId = doc.getString("flatId") ?: "",
                occupantEmail = doc.getString("OccupantEmail") ?: "",
                occupantName = doc.getString("OccupantName") ?: "",
                occupantId = doc.getString("OccupantId") ?: "",
                category = doc.getString("Category") ?: "",
                date = date,
                status = doc.getString("Status") ?: "OPEN",
                resolveDate = resolveDate,
                priority = doc.getString("Priority") ?: "",
                description = doc.getString("Description") ?: "",
                problem = doc.getString("Problem") ?: "",
                mediaUrls = (doc.get("MediaUrls") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                workerName = doc.getString("WorkerName") ?: "",
                workerUid = doc.getString("WorkerUid") ?: "",
                workerRemarks = doc.getString("WorkerRemarks") ?: "",
                workerMedia = (doc.get("WorkerMedia") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                userCompletionRemarks = doc.getString("UserCompletionRemarks") ?: "",
                adminCloseRemarks = doc.getString("AdminCloseRemarks") ?: ""
            )
        }
    }

    override suspend fun markCompletedByUser(complaintId: String, userRemarks: String) {
        val trimmed = userRemarks.trim().take(255)
        db.collection("Complaints").document(complaintId)
            .update(
                mapOf(
                    "Status" to "COMPLETED",
                    "UserCompletionRemarks" to trimmed,
                ),
            )
            .await()
    }

    override suspend fun closeComplaintByUser(complaintId: String, userRemarks: String) {
        val trimmed = userRemarks.trim().take(255)
        val updates = mutableMapOf<String, Any>(
            "Status" to "CLOSED",
            "ResolveDate" to java.text.SimpleDateFormat("yyyy-MM-dd", java.util.Locale.US)
                .format(java.util.Date()),
        )
        if (trimmed.isNotBlank()) {
            updates["UserCompletionRemarks"] = trimmed
        }
        db.collection("Complaints").document(complaintId)
            .update(updates)
            .await()
    }

    fun observeByOccupant(occupantId: String): Flow<List<Complaint>> = callbackFlow {
        val registration = db.collection("Complaints")
            .whereEqualTo("OccupantId", occupantId)
            .orderBy("Date", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null || snapshot == null) { trySend(emptyList()); return@addSnapshotListener }
                val complaints = snapshot.documents.map { doc ->
                    val date = (doc.get("Date") as? Timestamp)?.toDate()?.time ?: 0L
                    val rd = doc.get("ResolveDate")
                    val resolveDate = if (rd is Timestamp) rd.toDate().time else 0L
                    Complaint(
                        id = doc.id,
                        flatNumber = doc.getString("FlatNumber") ?: "",
                        flatId = doc.getString("flatId") ?: "",
                        occupantEmail = doc.getString("OccupantEmail") ?: "",
                        occupantName = doc.getString("OccupantName") ?: "",
                        occupantId = doc.getString("OccupantId") ?: "",
                        category = doc.getString("Category") ?: "",
                        date = date,
                        status = doc.getString("Status") ?: "OPEN",
                        resolveDate = resolveDate,
                        priority = doc.getString("Priority") ?: "",
                        description = doc.getString("Description") ?: "",
                        problem = doc.getString("Problem") ?: "",
                        mediaUrls = (doc.get("MediaUrls") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                        workerName = doc.getString("WorkerName") ?: "",
                        workerUid = doc.getString("WorkerUid") ?: "",
                        workerRemarks = doc.getString("WorkerRemarks") ?: "",
                        workerMedia = (doc.get("WorkerMedia") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                        userCompletionRemarks = doc.getString("UserCompletionRemarks") ?: "",
                        adminCloseRemarks = doc.getString("AdminCloseRemarks") ?: ""
                    )
                }
                trySend(complaints)
            }
        awaitClose { registration.remove() }
    }
}

File: iosApp/iosApp/BackendComplaintRepository.swift
import Foundation
import FirebaseFirestore

struct Complaint: Identifiable, Hashable {
    let id: String
    let flatNumber: String
    let flatId: String
    let category: String
    let date: Date
    let status: String
    let resolveDate: Date?
    let priority: String
    let description: String
    let problem: String
    let mediaUrls: [String]
    let workerName: String
    let workerRemarks: String
    let workerMedia: [String]
    let occupantId: String
    /// Optional note the occupant attaches when marking the complaint COMPLETED.
    let userCompletionRemarks: String
    /// Mandatory note the admin attaches when CLOSING the complaint.
    let adminCloseRemarks: String
}

class BackendComplaintRepository {
    private let db = Firestore.firestore()

    func submitComplaint(
        flatNumber: String,
        flatId: String,
        occupantEmail: String,
        occupantName: String,
        occupantId: String,
        category: String,
        priority: String,
        description: String,
        problem: String,
        mediaUrls: [String] = []
    ) async throws -> String {
        let data: [String: Any] = [
            "FlatNumber": flatNumber,
            "flatId": flatId,
            "OccupantEmail": occupantEmail,
            "OccupantName": occupantName,
            "OccupantId": occupantId,
            "Category": category,
            "Date": Timestamp(date: Date()),
            "Status": "OPEN",
            "ResolveDate": "",
            "Priority": priority,
            "Description": description,
            "Problem": problem,
            "MediaUrls": mediaUrls,
            "WorkerName": "",
            "WorkerUid": "",
            "WorkerRemarks": "",
            "WorkerMedia": [String]()
        ]
        let ref = try await db.collection("Complaints").addDocument(data: data)
        return ref.documentID
    }

    func fetchByOccupant(occupantId: String) async throws -> [Complaint] {
        let snapshot = try await db.collection("Complaints")
            .whereField("OccupantId", isEqualTo: occupantId)
            .order(by: "Date", descending: true)
            .getDocuments()
        return snapshot.documents.compactMap { doc -> Complaint? in
            let data = doc.data()
            guard let category = data["Category"] as? String else { return nil }
            let date = (data["Date"] as? Timestamp)?.dateValue() ?? Date()
            let resolveDate = (data["ResolveDate"] as? Timestamp)?.dateValue()
            return Complaint(
                id: doc.documentID,
                flatNumber: data["FlatNumber"] as? String ?? "",
                flatId: data["flatId"] as? String ?? "",
                category: category,
                date: date,
                status: data["Status"] as? String ?? "OPEN",
                resolveDate: resolveDate,
                priority: data["Priority"] as? String ?? "",
                description: data["Description"] as? String ?? "",
                problem: data["Problem"] as? String ?? "",
                mediaUrls: data["MediaUrls"] as? [String] ?? [],
                workerName: data["WorkerName"] as? String ?? "",
                workerRemarks: data["WorkerRemarks"] as? String ?? "",
                workerMedia: data["WorkerMedia"] as? [String] ?? [],
                occupantId: data["OccupantId"] as? String ?? "",
                userCompletionRemarks: data["UserCompletionRemarks"] as? String ?? "",
                adminCloseRemarks: data["AdminCloseRemarks"] as? String ?? ""
            )
        }
    }

    /// Occupant marks a complaint COMPLETED with an optional feedback note.
    /// Admin still has to CLOSE the complaint from the admin portal.
    func markCompletedByUser(id: String, userRemarks: String) async throws {
        let trimmed = String(userRemarks.trimmingCharacters(in: .whitespacesAndNewlines).prefix(255))
        try await db.collection("Complaints").document(id).updateData([
            "Status": "COMPLETED",
            "UserCompletionRemarks": trimmed,
        ])
    }

    /// Occupant CLOSES a complaint that the worker already marked COMPLETED.
    /// Optional feedback lands in the same `UserCompletionRemarks` field.
    func closeComplaintByUser(id: String, userRemarks: String) async throws {
        let trimmed = String(userRemarks.trimmingCharacters(in: .whitespacesAndNewlines).prefix(255))
        let fmt = DateFormatter()
        fmt.locale = Locale(identifier: "en_US_POSIX")
        fmt.dateFormat = "yyyy-MM-dd"
        var updates: [String: Any] = [
            "Status": "CLOSED",
            "ResolveDate": fmt.string(from: Date()),
        ]
        if !trimmed.isEmpty {
            updates["UserCompletionRemarks"] = trimmed
        }
        try await db.collection("Complaints").document(id).updateData(updates)
    }

    func observeByOccupant(occupantId: String, onChange: @escaping ([Complaint]) -> Void) -> ListenerRegistration {
        return db.collection("Complaints")
            .whereField("OccupantId", isEqualTo: occupantId)
            .order(by: "Date", descending: true)
            .addSnapshotListener { snapshot, _ in
                guard let snapshot = snapshot else { onChange([]); return }
                let complaints: [Complaint] = snapshot.documents.compactMap { doc in
                    let data = doc.data()
                    guard let category = data["Category"] as? String else { return nil }
                    let date = (data["Date"] as? Timestamp)?.dateValue() ?? Date()
                    let resolveDate = (data["ResolveDate"] as? Timestamp)?.dateValue()
                    return Complaint(
                        id: doc.documentID,
                        flatNumber: data["FlatNumber"] as? String ?? "",
                        flatId: data["flatId"] as? String ?? "",
                        category: category,
                        date: date,
                        status: data["Status"] as? String ?? "OPEN",
                        resolveDate: resolveDate,
                        priority: data["Priority"] as? String ?? "",
                        description: data["Description"] as? String ?? "",
                        problem: data["Problem"] as? String ?? "",
                        mediaUrls: data["MediaUrls"] as? [String] ?? [],
                        workerName: data["WorkerName"] as? String ?? "",
                        workerRemarks: data["WorkerRemarks"] as? String ?? "",
                        workerMedia: data["WorkerMedia"] as? [String] ?? [],
                        occupantId: data["OccupantId"] as? String ?? "",
                        userCompletionRemarks: data["UserCompletionRemarks"] as? String ?? "",
                        adminCloseRemarks: data["AdminCloseRemarks"] as? String ?? ""
                    )
                }
                onChange(complaints)
            }
    }
}
